ang/crmCaseType.js

```diff
@@ -460,31 +479,55 @@
-    // https://lab.civicrm.org/dev/core/issues/774
+    // Note that roleName passed to addRole is a misnomer, it's passed as the
+    // label, but we want to pass on as the internal name.
+    // Additionally, roleName is backwards from anything we have defined in
+    // $scope.caseType.definition.caseRoles, so we need to do some work to
+    // tell if it's already in the list, but it's work we have to do anyway
+    // to pass on the internal name.
 
     /// Add a new role
     $scope.addRole = function(roles, roleName) {
+      var matchingRole;
+      // does the label we've been given match up to any relationship type, based on id (which is label)
+      var matchingRoles = _.filter($scope.relationshipTypeOptions, {id: roleName});
+      if (matchingRoles.length) {
+        matchingRole = matchingRoles.shift();
+      }
+      // if found, is the corresponding machine name in the list of existing roles for the case type
       var names = _.pluck($scope.caseType.definition.caseRoles, 'name');
-      if (!_.contains(names, roleName)) {
-        var matchingRoles = _.filter($scope.relationshipTypeOptions, {id: roleName});
-        if (matchingRoles.length) {
-          var matchingRole = matchingRoles.shift();
-          roles.push({name: roleName, displaylabel: matchingRole.text});
-        } else {
-           CRM.loadForm(CRM.url('civicrm/admin/reltype', {action: 'add', reset: 1, label_a_b: roleName}))
-            .on('crmFormSuccess', function(e, data) {
-              var newType = _.values(data.relationshipType)[0];
-              roles.push({name: newType.label_b_a, displaylabel: newType.label_a_b});
-              // Assume that the case role should be A-B but add both directions as options.
-              $scope.relationshipTypeOptions.push({id: newType.label_a_b, text: newType.label_a_b});
-              if (newType.label_a_b != newType.label_b_a) {
-                $scope.relationshipTypeOptions.push({id: newType.label_b_a, text: newType.label_b_a});
-              }
-              $scope.$digest();
-            });
+      if (matchingRole) {
+        if (!_.contains(names, matchingRole.xmlNameOtherDirection)) {
+          roles.push({name: matchingRole.xmlNameOtherDirection, displaylabel: matchingRole.text});
         }
+      } else {
+         // not a known relationship type, so create on-the-fly
+         CRM.loadForm(CRM.url('civicrm/admin/reltype', {action: 'add', reset: 1, label_a_b: roleName}))
+          .on('crmFormSuccess', function(e, data) {
+            var newType = _.values(data.relationshipType)[0];
+            roles.push({name: newType.name_b_a, displaylabel: newType.label_a_b});
+            // Assume that the case role should be A-B but add both directions as options.
+            var newRelTypeOption = {
+              xmlNameOtherDirection: newType.name_b_a,
+              xmlName: newType.name_a_b,
+              id: newType.name_a_b,
+              text: newType.label_a_b
+            };
```

Comment:
This got the directions and the names a little mixed up.  The `xmlName` should be the same as `roles.name` above.  The `text` is correctly the same as `roles.displaylabel` above.  `id` should be the label counterpart to `xmlName`, and `xmlNameOtherDirection` should be the name counterpart to `text`.  The result should look like this:

```js
            var newRelTypeOption = {
              xmlName: newType.name_b_a,
              xmlNameOtherDirection: newType.name_a_b,
              id: newType.label_b_a,
              text: newType.label_a_b
            };
```